File: 346. Moving Average from Data Stream.cpp
class MovingAverage {
private:
    double sum;
    queue<int> cache;
    int window;
public:
    /** Initialize your data structure here. */
    MovingAverage(int size) {
        window = size;
        sum = 0;
    }
    // queue, O(1), 85.26%, ok
    double next(int val) {
        if(!window) return 0.0;
        cache.push(val);
        sum += val;
        if(cache.size() > window){
            sum -=cache.front();
            cache.pop();
        }
        return sum / cache.size();
    }
};

/**
 * Your MovingAverage object will be instantiated and called as such:
 * MovingAverage obj = new MovingAverage(size);
 * double param_1 = obj.next(val);
 */

File: 346. Moving Average from Data Stream.py
from collections import deque
class MovingAverage(object):

    # queue: O(1) time, O(size) space
    def __init__(self, size):
        """
        Initialize your data structure here.
        :type size: int
        """
        self.queue = deque([])
        self.size = size
        self.sum = 0
                
    def next(self, val):
        """
        :type val: int
        :rtype: float
        """
        if len(self.queue) < self.size:
            self.sum += val
        else:
            self.sum += val - self.queue.popleft()
        self.queue.append(val)
        return self.sum * 1.0 / len(self.queue)

    # queue, pythonic: O(size) time, O(size) space
    def __init__(self, size):
        self.queue = deque(maxlen=size)
                
    def next(self, val):
        self.queue.append(val)
        return float(sum(self.queue)) / len(self.queue)
        


# Your MovingAverage object will be instantiated and called as such:
# obj = MovingAverage(size)
# param_1 = obj.next(val)